docs/samples/turing-machine.mgn

```diff
@@ -0,0 +1,127 @@
+    (cons tape (cons head (cons state (cons program nil))))
+
+/** runs the given machine */
+def runMachine :=
+  Y (\rec.machine.
+     (eq (getState machine) 0)
+      machine
+      (rec (executeStep machine))
+  )
+
+/** executes one step of the given turing machine */
+def executeStep :=
+  \machine.
+     (eq (getState machine) 0)
+      machine
+      (nextMachine machine (getInstruction (getState machine)
+                                (getTapeContent machine)
+                                (getProgram machine)))
+
+/** creates a new machine based on the provided instruction */
+def nextMachine :=
+  \machine.instr.
+    (newMachine
+      (copyTapeWith (getTape machine) (getN 2 instr) (getHead machine))
+      ((eq (getN 3 instr) 0) (minus (getHead machine) 1) (plus (getHead machine) 1))
+      (getN 4 instr)
+      (getProgram machine)
+    )
+
+/** copies the given tape with the character at the specified index */
+def copyTapeWith :=
+  \tape.character.index.
+    Y (\rec.tape.counter.acc.
+       (isnil tape)
+        ((eq index counter)
+          (reverse (cons character acc))
+          ((gt index counter)
+            (rec nil (plus counter 1) (cons 0 acc))
+            (reverse acc)
+          )
+        )
+        (rec (tail tape)
+             (plus counter 1)
+             (cons ((ne counter index)
+               (head tape)
+               character
+             ) acc))
+    ) tape 0 nil
+
+
+/** returns the instruction for the given state */
+def getInstruction :=
+  \state.character.program.
+    Y (\rec.program.
+       (and (eq (head (head program)) state)
+            (eq (head (tail (head program))) character))
+              (head program)
+              (rec (tail program))
+    ) program
+
+/** returns the program of the given turing machine */
+def getProgram := \machine.head (applyN tail 3 machine)
+
+/** returns the head of the given turing machine */
+def getHead := \machine.head (tail machine)
+
+/** returns the state of the given turing machine */
+def getState := \machine.head (applyN tail 2 machine)
+
+/** returns the content of the tape at the current head position */
+def getTapeContent := \machine.getTapeContentAt (getTape machine) (getHead machine)
+
+/** returns the tape content at the specified index */
+def getTapeContentAt :=
+  Y (\rec.tape.index.
+     (isnil tape)
+      0
+      ((eq index 0)
+        (head tape)
+        (rec (tail tape) (minus index 1)))
+  )
+
+/** returns the tape of the given turing machine */
+def getTape := \machine.head machine
+
+/** returns the element at the n-th position */
+def getN := Y (\rec.index.list. (gt index 0) (rec (minus index 1) (tail list)) (head list))
+
+/**
+ * Example usage:
+ *
+ * def tape := []
+ * def program := [
+ *   [1, 0, 1, 1, 2],
+ *   [1, 2, 'x', 1, 0],
+ *   [2, 0, 2, 1, 3],
+ *   [3, 0, 1, 0, 1]
+ * ]
+ * def head := 0
+ * def start := 1
+ * def machine := newMachine tape head start program
+ * def main := runMachine machine 
+ */
```

Comment:
I'd suggest to organize the comments in this section like:

```c
/** Example usage: */
/*
def tape := []
def program := [
  [1, 0, 1, 1, 2],
  [1, 2, 'x', 1, 0],
  [2, 0, 2, 1, 3],
  [3, 0, 1, 0, 1]
]
def head := 0
def start := 1
def machine := newMachine tape head start program
def main := runMachine machine 
*/
```

It makes it easier to uncomment even in the most minimalistic text editor in the world.